Add Mikael's DevBlog (#3338)
dbt-labs/docs.getdbt.com: website/blog/2023-05-04-generating-dynamic-docs.md

---

> title: "Generating dynamic dbt docs for a whole domain at once"

Reviewer: Not sure about this title - dynamic to me implies that they're self-updating over time. This programmatically finds models and columns to document, but once they're documented there's no more dynamism.

> title: "Generating dynamic dbt docs for a whole domain at once"

Reviewer 2: fair point! I'm struggling a bit with a good title, and very open to suggestions :) Here's another suggestion:
```suggestion
title: "Accelerate your documentation workflow: Generate docs for whole folders at once"
```

> description: "For columns that are reused across models, Mikael walks through a DRY-method to generate dynamic documentation using the dbt Codegen package and docs blocks."

Reviewer 2: ```suggestion
description: "For columns that are reused across models, Mikael walks through a DRY-method to make documentation easier, using the dbt Codegen package and docs blocks."
```

slug: generating-dynamic-docs-dbt

authors: [mikael_thorup]

tags: [dbt tutorials]
hide_table_of_contents: false

date: 2023-05-09
is_featured: true
---

> At [Lunar](https://www.lunar.app/), most of our dbt models are sourcing from event-driven architecture. As an example, we have the following models for our account domain in our ingestion layer:

Reviewer: It'd be worth explaining providing a definition of "domain" (even just saying it's a folder full of models related to the same thing).

Reviewer: Alternatively, you could probably get broader understanding by just replacing the word "domain" with "folder" across the board.

> At [Lunar](https://www.lunar.app/), most of our dbt models are sourcing from event-driven architecture. As an example, we have the following models for our account domain in our ingestion layer:

Reviewer 2: ```suggestion
At [Lunar](https://www.lunar.app/), most of our dbt models are sourcing from event-driven architecture. As an example, we have the following models for our `activity_based_interest` folder in our ingestion layer:
```

- `account_created`
- `account_access_added` (for shared account access)
- `account_access_removed` (for shared account access)
- `user_primary_account`
- `account_closed`

> - `accountnumber_to_user` (who owns what accounts)

Reviewer: As far as I can see, this is the only time that these models are referenced. Given that the rest of the article talks about the `activity_based_interest` domain, it might make more sense to use that as your initial example as well so that it stays relevant

> - `account_created`
> - `account_access_added` (for shared account access)
> - `account_access_removed` (for shared account access)
> - `user_primary_account`
> - `account_closed`
> - `accountnumber_to_user` (who owns what accounts)

Reviewer 2: ```suggestion
- `activity_based_interest_activated.sql`
- `activity_based_interest_deactivated.sql`
- `activity_based_interest_updated.sql`
- `downgrade_interest_level_for_user.sql`
- `set_inactive_interest_rate_after_july_1st_in_bec_for_user.sql`
- `set_inactive_interest_rate_from_july_1st_in_bec_for_user.sql`
- `set_interest_levels_from_june_1st_in_bec_for_user.sql`
```

This results in a lot of the same columns (e.g. `account_id`) existing in different models, across different layers. This means I end up:

1. Writing/copy-pasting the same documentation over and over again
1. Halfway through, realizing I could improve the wording to make it easier to understand, and go back and update the `.yml` files I already did
1. Realizing I made a syntax error in my `.yml` file, so I go back and fix it

> 1. Realizing the columns are defined differently with different wording being used in other domains

Reviewer 2: ```suggestion
1. Realizing the columns are defined differently with different wording being used in other folders in our dbt project
```

1. Reconsidering my choice of career and pray that a large language model will steal my job
1. Considering if there’s a better way to be generating documentation used across different models

<!--truncate-->

> In fact, I found a better way, and made the following meme in the [dbt Community Slack](https://www.getdbt.com/community/join-the-community/) channel #memes-and-off-topic-chatter to encapsulate this method:

Reviewer: I think it's worth calling out right up top that you are using docs blocks - when there was no mention of them until very late in the piece, I spent most of my readthrough thinking you didn't know about them which made me worry that you had built something unnecessary. On the contrary, this is a cool way of using docs blocks more effectively

> In fact, I found a better way, and made the following meme in the [dbt Community Slack](https://www.getdbt.com/community/join-the-community/) channel #memes-and-off-topic-chatter to encapsulate this method:

Reviewer 2: ```suggestion
In fact, I found a better way using some CLI commands, the dbt Codegen package and docs blocks. I also made the following meme in the [dbt Community Slack](https://www.getdbt.com/community/join-the-community/) channel #memes-and-off-topic-chatter to encapsulate this method:
```

<Lightbox src="/img/blog/2023-05-04-generating-dynamic-docs/1.png" title="Meme of writing documentation" />

## What pain is being solved?

> If you build models on top of event driven architecture, this method limits manual errors, makes it faster to write and maintain documentation, and improves consistency of documentation. **This documentation method saves me 50-80% of the time I previously spent on documentation, by making the documentation process in dbt more <Term id="dry" /> and automated.**

Reviewer: >If you build models on top of event driven architecture

I think this is underselling your solution! It seems to me that it's useful _anywhere_ that you are documenting the same column multiple times

> If you build models on top of event driven architecture, this method limits manual errors, makes it faster to write and maintain documentation, and improves consistency of documentation. **This documentation method saves me 50-80% of the time I previously spent on documentation, by making the documentation process in dbt more <Term id="dry" /> and automated.**

Reviewer 2: <img width="250" alt="image" src="https://github.com/dbt-labs/docs.getdbt.com/assets/104904492/e67e6e45-910b-4bb6-af3e-3bd92079650a">

```suggestion
If you need to document the same column multiple times, this method limits manual errors, makes it faster to write and maintain documentation, and improves consistency of documentation. **This documentation method saves me 50-80% of the time I previously spent on documentation, by making the documentation process in dbt more <Term id="dry" /> and automated.**
```

```
13:16:21  Running with dbt=1.3.1
13:16:27  version: 2

models:
  - name: activity_based_interest_activated
    description: ""
    columns:
      - name: id
        description: ""

      - name: user_id
        description: ""

... (truncated for example purposes)

  - name: set_inactive_interest_rate_after_july_1st_in_bec_for_user
    description: ""
    columns:
      - name: id
        description: ""

      - name: user_id
        description: ""

      - name: start_date
        description: ""

      - name: event_time
        description: ""

      - name: event_day
        description: ""

  - name: set_inactive_interest_rate_from_july_1st_in_bec_for_user
    description: ""
    columns:
      - name: id
        description: ""

      - name: user_id
        description: ""

      - name: event_time
        description: ""

      - name: event_day
        description: ""
```

### Get model names programmatically

In order to not have to manually write all of the model names, we can programmatically gather names of relevant models:

```
$ dbt ls -m models/core/activity_based_interest --output name | xargs -I{} echo -n ' "{}",'

>  "activity_based_interest_activated", "activity_based_interest_deactivated", "activity_based_interest_updated", "downgrade_interest_level_for_user", "f_activity_based_interest", "set_inactive_interest_rate_after_july_1st_in_bec_for_user", "set_inactive_interest_rate_from_july_1st_in_bec_for_user", "set_interest_levels_from_june_1st_in_bec_for_user",%

Reviewer: This one could do with being split into two code blocks, to clarify that the second line is the output

Reviewer 2: Agree!

 ```

1. `dbt ls -m models/core/activity_based_interest`: This command lists all dbt models in the models/core/activity_based_interest directory.
1. `--output name`: This option filters the output to only show the name of each model, rather than the context + model name.
1. `| xargs -I{} echo -n ' "{}",'`: This pipe sends the output of the previous command to `xargs`, which runs the echo command on each line of output. 
    - `-I{}` specifies that `{}` should be replaced with the model name
    - The `echo` command then formats the model name by wrapping it in double quotes and appending a comma and a space: `"model", "name",` 
    - The `-n` option for `echo` removes the trailing newline

The output (⚠️ except the last two characters `,%` ) can then be copy-pasted into the following:

```
dbt run-operation generate_model_yaml --args '{"model_names": [ReplaceWithYourOutputFromPreviousCommand]}'
```

Which in turn can be copy-pasted into a new `.yml` file. In our example, we write it to `_activity_based_interest.yml`.

## Create docs blocks for the new columns

> [Docs blocks](https://docs.getdbt.com/docs/collaborate/documentation#using-docs-blocks) can be utilized to write more DRY and robust documentation. To use docs blocks normally, we could manually update our folder structure, the `.md` file, and the `.yml` file. Your file structure would look a bit like this:

Reviewer: >To use docs blocks normally, we could manually update our folder structure

This sounds like you're gearing up to describe what to do instead, but you want them to do this right? You should switch to an imperative voice (e.g. "To use docs blocks, update your folder structure to contain an `.md` file.")

> [Docs blocks](https://docs.getdbt.com/docs/collaborate/documentation#using-docs-blocks) can be utilized to write more DRY and robust documentation. To use docs blocks normally, we could manually update our folder structure, the `.md` file, and the `.yml` file. Your file structure would look a bit like this:

Reviewer 2: ```suggestion
[Docs blocks](https://docs.getdbt.com/docs/collaborate/documentation#using-docs-blocks) can be utilized to write more DRY and robust documentation. To use docs blocks, update your folder structure to contain a `.md` file. Your file structure should now look like this:
```

│   ├── set_inactive_interest_rate_from_july_1st_in_bec_for_user.sql
│   └── set_interest_levels_from_june_1st_in_bec_for_user.sql
└── models
    └── f_activity_based_interest.sql
```

```
$ cat models/core/activity_based_interest/_activity_based_interest_docs.md
{% docs activity_based_interest__id %}  

Primary key of the table. See sql for key definition.

{% enddocs %}

{% docs activity_based_interest__user_id %}  

The internal company id for a given user.

{% enddocs %}
```

```
$ cat models/core/activity_based_interest/_activity_based_interest_docs.yml
version: 2

models:
  - name: activity_based_interest_activated
    description: ""
    columns:
      - name: id
        description: "{{ doc('activity_based_interest__id') }}"

      - name: user_id
        description: "{{ doc('activity_based_interest__user_id') }}"

... (truncated for example purposes)

  - name: set_inactive_interest_rate_after_july_1st_in_bec_for_user
    description: ""
    columns:
      - name: id
        description: "{{ doc('activity_based_interest__id') }}"

      - name: user_id
        description: "{{ doc('activity_based_interest__user_id') }}"
```

> To confirm the formatting works, we run the following command to get dbt Docs up and running:

Reviewer 2: ```suggestion
To confirm the formatting works, run the following command to get dbt Docs up and running:
```

```
$ dbt docs && dbt docs serve
```
<Lightbox src="/img/blog/2023-05-04-generating-dynamic-docs/2.jpg" title="dbt Docs UI" />

> Here, we can confirm that the column descriptions using the doc blocks are correct.

Reviewer 2: ```suggestion
Here, you can confirm that the column descriptions using the doc blocks are working as intended.
```

 
:::note
Note that we would ideally like to use the `user_id` column definition from our `user_domain`, but for the sake of example, we’ll just go with it for now.

> :::

Reviewer: What does this mean?

> :::note
> Note that we would ideally like to use the `user_id` column definition from our `user_domain`, but for the sake of example, we’ll just go with it for now.
> :::

Reviewer 2: In our case, user_id is a foreign key to one of our user table, where we have a docs block for user_id. So we'd want to reuse that docs block. But I reckon this might be more confusing that really improve understanding, so perhaps just scrap that whole sentence. 
```suggestion
```

Reviewer 2: Also just realized this is already covered [here](https://github.com/dbt-labs/docs.getdbt.com/pull/3338/files/f987c1e3e96ed15cc406d3d9d0e6dc621a08568f#diff-26655aab68fa333d27ea3acf6bf69c6c3b1a4aea2f32ad67bac790d70a2ce353R338), so this note here could definitely be deleted

> ### Get all columns within the domain

Reviewer 2: ```suggestion
### Get all unique columns within the folder
```

> To cut down on copy-pasting between your markdown and YAML files, we want to find all of the unique columns in the domain, and run the following:

Reviewer: This is a place where I got confused on my first read through, because I wasn't clear on what the "domain" was.

> To cut down on copy-pasting between your markdown and YAML files, we want to find all of the unique columns in the domain, and run the following:

Reviewer 2: ```suggestion
To cut down on copy-pasting between your markdown and YAML files, find all of the unique columns in the folder and subfolders, by running the following command:
```

```
$ grep '      \- name:' models/core/activity_based_interest/_activity_based_interest_docs.yml | cut -c 15- | sort -u
end_date
event_day
event_time
id
is_active
last_updated_date
start_date
tier_interest_percentage
tier_threshold_amount
user_id
```

Breaking down this command:
- `grep ' \- name:' models/core/activity_based_interest/_activity_based_interest_docs.yml` searches for the pattern ` - name:` in the file `_activity_based_interest_docs.yml` located in the directory `models/core/activity_based_interest/`.

> - `cut -c 15-` removes the first 14 characters of each line from the output.

Reviewer: To make this more generally applicable, it would be useful to know how you chose `14` as the number of characters to cut. I spent a bit of time looking at earlier files in the blog post and couldn't work out where you were coming from

> - `cut -c 15-` removes the first 14 characters of each line from the output.

Reviewer 2: hehe yeah true - it is pretty confusing. 
```suggestion
- `cut -c 15-` cuts the first 14 characters of each line from the output, i.e. in .yml files, we cut `      - name: ` from `      - name: some_column_name`, so you are left with only `some_column_name`.
```

- `sort -u` sorts the output in alphabetical order and removes any duplicate lines.

### Format to align with Jinja docs block

Copy-paste the above output into your `.md` file, so it looks like the following: 

```
$ cat models/core/activity_based_interest/_activity_based_interest_docs.md
end_date
event_day
event_time
id
is_active
last_updated_date
start_date
tier_interest_percentage
tier_threshold_amount
user_id
```
Now, open your code editor, and replace `(.*)` with `{% docs column__activity_based_interest__$1 %}\n\n{% enddocs %}\n`, which will result in the following in your markdown file:

<Lightbox src="/img/blog/2023-05-04-generating-dynamic-docs/3.png" title="Replace content in your markdown file" />

Now you can add documentation to each of your columns.

## Update `.yml` file to source documentation from the `.md` file

> We can programmatically identify all columns, and have them point towards the newly-created documentation. In your code editor, replace `\s{6}- name: (.*)\n        description: ""` with `      - name: $1\n        description: "{{ doc('column__activity_based_interest__$1') }}`:

Reviewer 2: ```suggestion
You can programmatically identify all columns, and have them point towards the newly-created documentation. In your code editor, replace `\s{6}- name: (.*)\n        description: ""` with `      - name: $1\n        description: "{{ doc('column__activity_based_interest__$1') }}`:
```

<Lightbox src="/img/blog/2023-05-04-generating-dynamic-docs/4.png" title="Replace descriptions with dynamic doc blocks" />

> ⚠️ Some of your columns may have already been documented in other domains. In our example, we want to make the following replacements:

Reviewer 2: ```suggestion
⚠️ Some of your columns may already be available in existing docs blocks. In this example, the following replacements are done:
```

- `{{ doc('column__activity_based_interest__user_id') }}` → `{{ doc("column_user_id") }}`
- `{{ doc('column__activity_based_interest__event_day') }}` → `{{ doc("column_event_day") }}`

## Check that everything works

> Now we run `dbt docs generate`. If there are syntax errors, this will be found out at this stage. If successful, we can run `dbt docs serve` to perform a smoke test and ensure everything looks right:

Reviewer 2: ```suggestion
Run `dbt docs generate`. If there are syntax errors, this will be found out at this stage. If successful, we can run `dbt docs serve` to perform a smoke test and ensure everything looks right:
```

<Lightbox src="/img/blog/2023-05-04-generating-dynamic-docs/5.jpg" title="dbt Docs UI showing successful documentation leveraging docs blocks" />

## Additional considerations

- Q: What about when they are slight deviations in column documentation between models?
    - A: I’ve been using dynamic documentation to contain the "essence" of the documentation, and then appending static documentation, like so:

    ```yaml
           - name: user_id
             description: "{{ doc('dynamic_docs') }}, additional static info" 
    ```

> - Q: Should I use this approach on modifications to an existing domain? 
>     - A: When adding additional models, or additional columns to an existing domain, I would suggest adding new documentation manually.

Reviewer 2: ```suggestion
- Q: Should I use this approach on modifications to an existing folder? 
    - A: When adding additional models to a folder, or additional columns to an existing model, I would suggest adding new documentation and docs blocks manually rather than programmatically.
```

---
title: "Accelerate your documentation workflow: Generate docs for whole folders at once"
description: "For columns that are reused across models, Mikael walks through a DRY-method to make documentation easier, using the dbt Codegen package and docs blocks."

> slug: generating-dynamic-docs-dbt

Reviewer: Something like bulk-docs-generation? @KiraFuruichi I assume you'll have specific ideas on a search-friendly slug
